les/clientpool.go

```diff
@@ -559,75 +851,87 @@ func (f *clientPool) addBalance(id enode.ID, amount int64, meta string) (uint64,
 	f.lock.Lock()
 	defer f.lock.Unlock()
 
+	now := f.clock.Now()
 	pb := f.ndb.getOrNewPB(id)
-	var negBalance uint64
+	var negBalance expiredValue
 	c := f.connectedMap[id]
 	if c != nil {
-		pb.value, negBalance = c.balanceTracker.getBalance(f.clock.Now())
+		pb.value, negBalance = c.balanceTracker.getBalance(now)
 	}
 	oldBalance := pb.value
-	if amount > 0 {
-		if amount > maxBalance || pb.value > maxBalance-uint64(amount) {
-			return oldBalance, oldBalance, errBalanceOverflow
-		}
-		pb.value += uint64(amount)
-	} else {
-		if uint64(-amount) > pb.value {
-			pb.value = 0
-		} else {
-			pb.value -= uint64(-amount)
-		}
+	posExp := f.posExpiration(now)
+	oldValue := oldBalance.value(posExp)
+	if amount > 0 && (amount > maxBalance || oldValue > maxBalance-uint64(amount)) {
+		return oldValue, oldValue, errBalanceOverflow
 	}
+	pb.value.add(amount, posExp)
 	pb.meta = meta
 	f.ndb.setPB(id, pb)
 	if c != nil {
 		c.balanceTracker.setBalance(pb.value, negBalance)
-		if !c.priority && pb.value > 0 {
-			// The capacity should be adjusted based on the requirement,
-			// but we have no idea about the new capacity, need a second
-			// call to udpate it.
+		if c.active {
+			f.activeQueue.Update(c.queueIndex)
+			if !c.priority && pb.value.base > 0 {
+				// The capacity should be adjusted based on the requirement,
+				// but we have no idea about the new capacity, need a second
+				// call to udpate it.
+				f.priorityActive += c.capacity
+				f.updateFreeRatio()
+				c.balanceTracker.addCallback(balanceCallbackZero, 0, func() { f.balanceExhausted(id) })
+			}
+			c.balanceMetaInfo = meta
+			f.activeBalances.subExp(oldBalance)
+			f.activeBalances.addExp(pb.value)
+		} else {
+			f.inactiveQueue.Remove(c.queueIndex)
+			f.inactiveQueue.Push(c, -connPriority(c, f.clock.Now()))
+			f.inactiveBalances.subExp(oldBalance)
+			f.inactiveBalances.addExp(pb.value)
+		}
+		if pb.value.base > 0 {
 			c.priority = true
-			f.priorityConnected += c.capacity
-			c.balanceTracker.addCallback(balanceCallbackZero, 0, func() { f.balanceExhausted(id) })
+			// if balance is set to zero then reverting to non-priority status
+			// is handled by the balanceExhausted callback
 		}
-		// if balance is set to zero then reverting to non-priority status
-		// is handled by the balanceExhausted callback
-		c.balanceMetaInfo = meta
+	} else {
+		f.inactiveBalances.subExp(oldBalance)
+		f.inactiveBalances.addExp(pb.value)
 	}
-	return oldBalance, pb.value, nil
+	f.tryActivateClients()
+	return oldValue, pb.value.value(posExp), nil
 }
 
 // posBalance represents a recently accessed positive balance entry
 type posBalance struct {
-	value uint64
+	value expiredValue
```

Comment:
I would propose to separate the pcBalance out.

What about this approach:
- Rename `posBalance` and `negBalance` to `posServiceToken` and `negServiceToken`.
- Define pcBalance as struct 
   ```go
    type pcBalance struct {
	    Name   string
	    Amount uint64
    }
   ```

- Define balanceList as []pcBalance
- Move all defintions and nodeDB stuff to a separate file `clientdb.go`

Reply:
Btw now the postive balance and negative balance should be exactly same. Why they have different definition?

Reply:
With the current token sale logic we cannot (and don't need to) handle different currency types here. The payment receiver can of course still accept multiple currencies (for example, ETH and DAI). The "pc" means "preferred currency" and it is just a technical thing. A payment receiver that accepts both ETH and DAI can convert all received funds to either ETH or DAI (I mean not necessarily exchange the actual money right away, just convert the amount for accounting purposes). Then the token sale mechanism allows conversion between the relatively stable `pcBalance` and the expiring and supply limited `serviceToken` at a dynamic rate depending on the issued amount and total limit. This mechanism works with two currencies, with more than two it would be a lot more complex so I think it is better to handle multiple currencies and do the conversion in an upper layer.

Reply:
About the other question: yes, they could be the same (they are actually the same in `balanceTracker` when they are connected). Negative balances are never manipulated though when not connected, so the original logarithmic version works fine too. I wanted to replace it actually, but then I noticed how much simpler the eviction condition in `nbEvictCallBack` is if I leave it in its original form :) From a readability perspective maybe it is better to change though so if you think I should change it to `expiredValue` then I will.
There is still one difference in the stored format though: negative balances do not need the `meta` field because there are no incoming payments to that balance.

Reply:
@zsfelfoldi Yes I agree. Too many PC types just make the whole over-complicated. My point is now it's complicated for the balance representation. Each client has 3 different types of balance.

So I think it would be super nice and clean if we can somehow uniform to a single structure `balance`. Something like this:

```go 

type balance struct {
	pos      expiredValue
	neg      expiredValue
	pc         uint64
	pctype string 
}

```
Whenever the neg balance is too small and no pos balance is assigned, we can drop the database entry.

And possibly there is an attack vector that client assigns itself a very small pos balance like 1 wei, so the database entry will never be deleted. So we can just set a minimal deposit amount to protect it. 

Anyway, just suggest to cleanup.

Reply:
Currently pc balance is stored together with the positive balance (it is encoded in the meta field associated with the service token balance). This is because the token sale method is an additional layer over the basic priority system which can be used without the token sale if the server operator has its own business model. I am not sure whether this is the best possible separation of layers but based on the previous feedback I received it seemed to me that it is a good idea to allow experimenting with different client priority models. We can debate it further, just saying that this is the reason it is currently implemented this way.
Storing the negative balances together with the rest is a bit more tricky because they are associated to IP address, not the node key because changing the node key is practically free. So I think they should be separate db entries.